security/sensorhub-security-oauth/src/main/java/org/sensorhub/impl/security/oauth/OAuthAuthenticator.java

```diff
@@ -235,30 +250,52 @@ public Authentication validateRequest(ServletRequest req, ServletResponse resp,
-                            log.debug("Client Credentials Token = {}", clientCredentialsToken);
+
+                            log.debug("Client Credentials Token: {}", clientCredentialsToken);
 
                             String[] chunks = clientCredentialsToken.split("\\.");
 
+                            // In validation we do not need to decode the entire token, just the header to perform validation
+                            // the userInfo will be extracted further below as we assign this token if valid as the accessToken
+                            // and idToken
                             Base64.Decoder decoder = Base64.getUrlDecoder();
 
                             String header = new String(decoder.decode(chunks[0]));
 
-                            log.debug("Client Credentials Token Header = {}", header);
-
                             JsonElement headerData = JsonParser.parseString(header);
 
-                            JwkProvider jwkProvider =
-                                    new JwkProviderBuilder(new URL(config.bearerTokenConfig.jwksUri))
-                                            .cached(config.bearerTokenConfig.cacheSize,
-                                                    Duration.of(config.bearerTokenConfig.cacheDuration, ChronoUnit.MINUTES))
-                                            .build();
-
-                            DecodedJWT decodedJWT;
                             try {
 
-                                RSAPublicKey publicKey = (RSAPublicKey) jwkProvider.get(parseKidFromJson(headerData)).getPublicKey();
+                                String kid = parseKidFromJson(headerData);
+
+                                Jwk jwk = null;
+
+                                try {
+                                    // Attempt to retrieve the JWK from the cache
+                                    jwk = jwkProvider.get(kid);
+
+                                } catch (JwkException e) {
+
+                                  log.info("JWK provider returned an error: {}", e.getMessage());
+                                }
+
+                                // If it does not exist, it is because it has expired and needs to be reset
+                                if (jwk == null) {
+
+                                    log.info("Creating a new JWK provider that will expire in {} minutes", config.bearerTokenConfig.cacheDuration);
+
+                                    this.jwkProvider =
+                                            new JwkProviderBuilder(new URL(config.bearerTokenConfig.jwksUri))
+                                                    .cached(config.bearerTokenConfig.cacheSize,
+                                                            Duration.of(config.bearerTokenConfig.cacheDuration, ChronoUnit.MINUTES))
+                                                    .build();
+
+                                    jwk = jwkProvider.get(kid);
+                                }
```

Comment:
Is recreating the provider really needed here?
According to the Javadoc, an exception on calling get(kid) is only thrown if it cannot find the kid at all.
So I thought that, if there is a cache miss, the provider would automatically refetch the key from the server.

Reply:
I was digging through the code for cache and I did not see it refetch the key from the server.  Leaving me to believe that the provider needed to be recreated in these cases.  I will dig into it some more to see if it is truly necessary to recreate the provider or if it will refetch automatically.

Reply:
I looked it over again, and I see where the flow will actually refetch the token.  I have simplified code by removing unnecessary logic.